### PR TITLE
net/acl: only allocate memory for used NUMA nodes

### PR DESCRIPTION
**Pull request by cjdoucette**
The memory pools and ACLs per NUMA are always allocated, even when a given NUMA node is not being used. This patch saves memory by only allocating memory pools and ACLs on NUMA nodes that are used, so that memory can be used by other critical data structures such as the flow table.

### CODE REVIEW

File: lib/net.c [hunk 1]
```diff
@@ -1085,6 +1085,63 @@ start_iface(struct gatekeeper_if *iface)
 	return ret;
 }
 
+static int
+init_net_stage1(void *arg)
```
Comment (mengxiang0811) [inline lib/net.c]: Shouldn't we allocate the memory pool on the fly? That is to say, when we encounter a new ``numa`` node, we allocate the corresponding memory pool. Since in the configuration file ``lua/gatekeeper_config.lua``, we will allocate ``lcores`` to different blocks on the fly. Initially, the ``numa_used[x]`` array should be all ``false``, the initialization below won't do anything.

Comment (cjdoucette) [reply lib/net.c]: I don't think all of the elements should be `false`, as the order of events should be:

(1) `lua/gatekeeper_config.lua` invokes the network configuration, which calls `gatekeeper_init_network()` to allocate and initialize `numa_used` to be an array of all `false`.
(2) `lua/gatekeeper_config.lua` invokes the individual block configurations, which allocate lcores. Based on the lcores that are allocated, the `numa_used` array is set to true for the corresponding sockets for those lcores.
(3) Stage 1 begins, and the first function that is invoked in this stage is now `init_net_stage1()`, which allocates memory pools for only the NUMA nodes that were set to `true` in `numa_used` by the individual block configurations.

**Comment by cjdoucette** [comment]
Just rebased with Qiaobin's suggestions and the latest commits on the master branch.

File: lib/net.c (hunk 1 above)
Comment (mengxiang0811) [reply lib/net.c]: You are right! It's launched at the first stage right after the Lua configuration finishes.

File: lib/acl.c [hunk 2]
```diff
@@ -136,11 +136,13 @@ process_acl(struct gatekeeper_if *iface, unsigned int lcore_id,
...
 destroy_acls(struct acl_state *astate)
 {
-	unsigned int numa_nodes = get_net_conf()->numa_nodes;
+	struct net_config *net_conf = get_net_conf();
 	unsigned int i;
-	for (i = 0; i < numa_nodes; i++) {
-		rte_acl_free(astate->acls[i]);
-		astate->acls[i] = NULL;
+	for (i = 0; i < net_conf->numa_nodes; i++) {
+		if (net_conf->numa_used[i]) {
```
Comment (AltraMayor) [inline lib/acl.c]: It would be simpler to just test `astate->acls[i] != NULL`. Analogous suggestion applies to other ACL functions such as `register_ipv4_acl()`, `build_ipv4_acls()`, etc. but `init_ipv4_acls()` and `init_ipv6_acls()`.

**Comment by cjdoucette** [comment]
This patch is ready for another review.

File: lib/acl.c [hunk 3]
```diff
@@ -136,11 +136,13 @@ process_acl(struct gatekeeper_if *iface, unsigned int lcore_id,
 void
 destroy_acls(struct acl_state *astate)
 {
-	unsigned int numa_nodes = get_net_conf()->numa_nodes;
+	struct net_config *net_conf = get_net_conf();
```
Comment (AltraMayor) [inline lib/acl.c]: This change made sense before the previous review, but now it does not since it's just looking up the field `net_conf->numa_nodes` at the end of the `for` loop instead of using variable `numa_nodes`.

This comment applies to a number of changes in this file, namely `lib/acl.c`.

**Comment by cjdoucette** [comment]
Just rebased with the suggestion.

File: lib/net.c [hunk 4]
```diff
@@ -1119,6 +1119,63 @@ start_iface(struct gatekeeper_if *iface)
...
+{
+	struct net_config *net_conf = arg;
+	uint32_t i;
+
+	if (net_conf->gatekeeper_pktmbuf_pool == NULL) {
+		net_conf->gatekeeper_pktmbuf_pool =
```
Comment (AltraMayor) [inline lib/net.c]: Given that we will now initialize the memory pools in stage 1 instead of during a call to `gatekeeper_init_network()`, we no longer need to test if `net_conf->gatekeeper_pktmbuf_pool == NULL`, that is, we can drop the test and unindent the body of the `if` statement.

File: lib/net.c [hunk 5]
```diff
@@ -1119,6 +1119,63 @@ start_iface(struct gatekeeper_if *iface)
...
+		}
+	}
+
+	/* Initialize pktmbuf pool on each used NUMA node. */
+	for (i = 0; i < net_conf->numa_nodes; i++) {
+		char pool_name[64];
+		int ret;
+
+		if (!net_conf->numa_used[i] ||
+				net_conf->gatekeeper_pktmbuf_pool[i] != NULL)
```
Comment (AltraMayor) [inline lib/net.c]: As explained in the previous comment, we will now initialize the memory pools in stage 1 instead of during a call to `gatekeeper_init_network()`, so we can simplify the test above to `!net_conf->numa_used[i]`.

File: lib/net.c [hunk 6]
```diff
@@ -1214,63 +1269,32 @@ gatekeeper_init_network(struct net_config *net_conf)
 	rte_convert_rss_key((uint32_t *)&default_rss_key,
 		(uint32_t *)rss_key_be, RTE_DIM(default_rss_key));
 
-	/* Initialize pktmbuf pool on each numa node. */
```
Comment (AltraMayor) [inline lib/net.c]: Bug: I can't comment on the line `if (randomize_rss_key() < 0) {` above, but that block cannot return `-1`, it has to `goto numa;`.

File: lib/net.c (hunk 4 above)
Comment (AltraMayor) [reply lib/net.c]: Discard this comment because we are not cleaning up if a failure occurs.

File: lib/net.c (hunk 5 above)
Comment (AltraMayor) [reply lib/net.c]: Discard this comment because we are not cleaning up if a failure occurs.
